File: Exercise/Hard/04. Bracket Validator/Code/Bracket Validator.js
function isValidExpression(expression) {
    const stack = [];
    const brackets = {
        "(": ")",
        "{": "}",
        "[": "]"
    };

    for (let char of expression) {
        if (brackets[char]) {
            stack.push(char);
        } else {
            const last = stack.pop();
            if (brackets[last] !== char) {
                return false;
            }
        }
    }

    return stack.length === 0;
}

let expression = "(({[({})]}))";
if (isValidExpression(expression)) {
    console.log("True");
} else {
    console.log("False");
}

File: Exercise/Hard/04. Bracket Validator/Code/Bracket Validator.cpp
#include <iostream>
#include <stack>
#include <string>

using namespace std;

bool isOpeningBracket(char ch) {
    return (ch == '(' || ch == '{' || ch == '[');
}

bool isValidPair(char opening, char closing) {
    return (opening == '(' && closing == ')') ||
           (opening == '{' && closing == '}') ||
           (opening == '[' && closing == ']');
}

bool isValidExpression(const string& expression) {
    stack<char> stack;

    for (char ch : expression) {
        if (isOpeningBracket(ch)) {
            stack.push(ch);
        } 
        else {
            if (stack.empty() || !isValidPair(stack.top(), ch)) {
                return false;
            }
            stack.pop();
        }
    }

    return stack.empty();
}

int main() {
    string expression;

    getline(cin, expression);

    if (isValidExpression(expression)) {
        cout << "True" << endl;
    } 
    else {
        cout << "False" << endl;
    }

    return 0;
}
